network/handlers/drain.go

```diff
@@ -115,19 +114,42 @@ func (d *Drainer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 func (d *Drainer) Drain() {
 	// Note: until the first caller exits, the others
 	// will wait blocked as well.
-	d.once.Do(func() {
-		t := func() timer {
-			d.Lock()
-			defer d.Unlock()
-			if d.QuietPeriod <= 0 {
-				d.QuietPeriod = network.DefaultDrainTimeout
+	ch := func() chan struct{} {
+		d.Lock()
+		defer d.Unlock()
+		if d.ch != nil {
+			return d.ch
+		}
+
+		if d.QuietPeriod <= 0 {
+			d.QuietPeriod = network.DefaultDrainTimeout
+		}
+
+		timer := newTimer(d.QuietPeriod)
+		ch := make(chan struct{})
+
+		go func() {
+			select {
+			case <-ch:
+				// closed by reset
+			case <-timer.tickChan():
+				close(ch)
 			}
-			d.timer = newTimer(d.QuietPeriod)
-			return d.timer
 		}()
 
-		<-t.tickChan()
-	})
+		d.ch = ch
+		d.timer = timer
+		return ch
+	}()
+
+	<-ch
+}
+
+func drainTimer(tc <-chan time.Time) {
```

Comment:
I don't get this, is it just resetting the timer when a signal comes on the channel?

Reply:
this drains the timer's channel - if the `Stop` call didn't stop the timer there may be an event here and we drain it.